docs/local_assembly.tex

```diff
@@ -0,0 +1,77 @@
+Read error correction is turned off by default in \HC~ and \Mutect~ and we are not familiar with it.  The idea is as follows: large kmers are good because they contain more phasing information and therefore yield a simpler de Bruijn graph.  For example, two SNVs 20 bases apart with sequenced with error-free reads yield a de Bruijn graph with $2 \times 2 = 4$ paths if $k = 10$ but only $2$ paths if $k = 40$, because the latter spans the phased SNVs.  The drawback is that larger kmers are more liable to be lost due to sequencing error simply because they contain more bases.  The idea of read error correction is to rescue kmers from the occasional error.
+
+First, the \code{ReadErrorCorrector} kmerizes every read and counts the occurrences of each kmer.  Then it builds a map of kmer corrections where kmers that appear often\footnote{By default, 20 times or more.  This threshold is set by the \code{minObservationsForKmerToBeSolid} command line argument.} map to themselves and kmers that appear only once\footnote{This is a hardcoded threshold.} map to their nearest neighbor in Hamming distance\footnote{For simplicity we consider only substitution errors, eg the distance between ACGGT and AGGTG is 3} within a maximum of two mismatches\footnote{This is a hardcoded threshold.}.  Kmers that appear between 1 and 20 times are not part of the correction map.
+
+Then, for every base in every read, the \code{ReadErrorCorrector} queries the kmer correction map for each overlapping kmer.  For example, to correct the fourth base of read ACGTATTC if k = 3, it looks at the corrections for CGT, GTA, and TAT at their thirds, second, and first positions, respectively.  If and only if the corrections are unanimous, the base is corrected.  Note that corrected reads are used only for assembly and the original reads are used downstream.
+
+\section{Building the graph} \label{graph-assembly}
+Next, the \code{ReadThreadingAssembler} assembles the (corrected) reads over several different kmer sizes specified by the \code{kmerSize} argument\footnote{By default, 10 and 25.  This size is a compromise and no single value is the best choice for all regions.  Large kmers are more likely to be unique and to yield a graph with no cycles, which is especially important in low-complexity regions, but they are more sensitive to errors and low coverage.}.  If the given kmer sizes fail to produce a graph without cycles, the \code{ReadThreadingAssembler} repeatedly increases $k$ by 10 bases until assembly succeeds.
+
+The first step is to create a set of sequences to be kmerized and put into the de Bruijn graph.  These sequences are: the reference haplotype, any requested alleles in \HC's GGA mode, and maximal subsequences of reads with base quality at least 10 and at least one kmer long.  That is, a read with 100 bases and a base of quality 7 at position 70 yields sequences of length 69 and 30 if $k = 10$ and a single sequence of length 69 if $k = 40$.
```

Comment:
> base quality at least 10 can be changed to base quality at least 10 by default because the default value of --min_base_quality_score is 10.

Done.

> 69, 30 would be clearer and read better.

Done

> Is this referring to requesting alleles with the --alleles argument?

Yes, and clarified.